### DISCUSSION TITLE
Assigning values (manually) to DOF

### DISCUSSION THREAD

**Discussion by federicodalinger** [kinnala/scikit-fem#1065] [Q&A]
Hi @kinnala, thanks again for answering my previous question.

I just wonder if there is a way to assign manually certain values for the DOF.

I am continuing with my project about plate bending. In this case, I am trying to calculate the deflection of a stiffened plate (considering water pressure on the other side of the plate). My mesh is like this:

![image](https://github.com/kinnala/scikit-fem/assets/74275861/3387c98e-ede3-420b-b036-5ad1e4dfff1b)

I would like to add the deflection, from a list of values, according to the typical beam bending formulas, on the middle horizontal blue line. In other words, I don't want to consider the displacement there as zero as per default, (although it will be really to it) because I wanna link those displacements to the geometry of that "flat beam".

I know I have to consider the "x=u" used in the ".condense" function, but, how could I add those displacements only for that middle blue line (it already has a defined name)?

Thanks in advance and kind regards,
Federico

**Comment by kinnala** [comment]
That blue line then becomes similar to a boundary where you assign a boundary condition, using condense like you have figured out.

Assuming you use C1 elements, each of those nodes on the blue line has several DOFs describing the displacement, first derivative and perhaps even second derivative. You also probably have a location-dependent value for the displacement. For this reason, it is probably not as simple as setting DOFs to some values.

What you need to do is perform L2 projection of the displacement field onto the finite element basis (over that blue segment). For this purpose, you most likely need to initialize InteriorFacetBasis object with facets='name_of_interface'. After this you use InteriorFacetBasis.project method.

However, you did not specify accurately enough in which format your input is. Is it a list of displacements only? At which points?

**Reply by federicodalinger** [reply]
Hi @kinnala, thanks so much for your quick explanation. Please see below a picture of what I have defined as the mesh and the basis for it:

![image](https://github.com/kinnala/scikit-fem/assets/74275861/b878f235-cbad-438a-84c3-8e2798a6e1b0)

Then, according to what you said (please, correct me if I'm wrong), I have:

![image](https://github.com/kinnala/scikit-fem/assets/74275861/ee5fce49-ca57-4c9e-8670-c55c7f2f4fbc)

Eventually, when I have to do the projection, I believe I have to do this:

![image](https://github.com/kinnala/scikit-fem/assets/74275861/cd40c8f7-50b7-4d9a-8f95-f9e076ce8b73)

That "internal_braces_displacement" is a function that will calculate the displacement (normal to the door sheet) in a way such a list of points (let's say 100 of them), which the code (not finished) is:

![image](https://github.com/kinnala/scikit-fem/assets/74275861/ea8e9978-632d-4410-aaa3-89a706190b3c)

So yes, to answer your question, the displacements I want to set on that "middle horizontal blue line", which is a flat plate, are the displacements obtained using the typical beam theory deflection formula (for a simply supported one by now, but later I will modify it to clamped because there will be 2 vertical braces close to the right and left sides).

I also think that I need to first find the "x" coordinates of each point that belongs to the blue line (the nodes), generate a list of points of those positions, and later, calculate the deflection on them. After doing this, I will have a proper displacement for the existing nodes.
Please, could you give me a hand on how to do it and perform the projection only for the displacement field? I would really appreciate ir.

Thanks in advance and wish you a lovely weekend,
Federico

**Reply by kinnala** [reply]
The code you have described should be correct or at least close to being correct. The method "project" will pass into internal_braces_displacement an array of integration points. Then you are expected to return the value of the function at those integration points. As a result you should have "u" which is a DOF array with values from the analytical beam solution. This can be passed to condense(..., x=u, D={'mid_brace_1', 'left', ...}).

**Reply by federicodalinger** [reply]
Hi @kinnala , thanks for your patience and for your replies. I finally found what my mistake was (although, as the Inertia is huge in comparison with the thin plate, it's almost hard to detect the changes.

I want to add here if you allow me a new question:
Is it possible to know the displacement value on one node, driven by the X and the Y coordinates? Please see the below image:

![image](https://github.com/kinnala/scikit-fem/assets/74275861/140d1f98-2915-41a2-a737-e39dfa2c8ec6)

The displacement comes from:

![image](https://github.com/kinnala/scikit-fem/assets/74275861/2322999f-16e4-4b85-9c41-c0a84c95aed8)

Thanks in advance, and again, I really appreciate the help you are giving me!

Kind regards,
Federico

**Reply by kinnala** [reply]
All the information is in the array 'w' and you just need to find the correct index.

To find the correct indices to 'w', you can use something like Basis.get_dofs(nodes=[(0.1, 0.1)]).all('u').

**Reply by federicodalinger** [reply]
Hi @kinnala, I hope you are well. 
Thanks mate, I have figured it out with your comment.
I really appreciate you for answering all my questions. My humble plate calculator is almost done!

Kind regards, 
Federico
